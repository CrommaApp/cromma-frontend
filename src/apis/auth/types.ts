import { BasicResult } from '@apis/index';

type userInfo = {
	Keywords: string[];
	createdAt: string;
	id: number;
	updatedAt: string;
	userId: string;
};

export type SignRequest = {
	userId: string;
	password: string;
};
export type LoginResult = BasicResult & {
	data: userInfo;
};

export type getMyInfoResult = BasicResult & {
	data: userInfo | null;
};
