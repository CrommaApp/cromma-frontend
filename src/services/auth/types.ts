import { BasicResult } from '@services/api';
import { Post } from '@services/post/types';

type userInfo = {
	createdAt: string;
	id: number;
	updatedAt: string;
	userId: string;
};

export type SignRequest = {
	userId: string;
	password: string;
};
export type LoginResult = BasicResult & {
	data: userInfo;
};

export type getMyInfoResult = BasicResult & {
	data: userInfo | null;
};
