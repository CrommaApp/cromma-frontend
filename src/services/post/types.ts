import { BasicResult } from '@services/api';

export type Post = {
	id: number;
	title: string;
	content: string;
	createdAt: string;
	updatedAt: string;
};

export type GetAllPostsResult = BasicResult & {
	data: Post[] & { UserId: string };
};

export type UploadPostRequest = {
	title: string;
	content: string;
};

export type UploadPostResult = BasicResult & {
	data: Post;
};

export type getPostRequest = {
	postId: string;
};

export type deletePostRequest = {
	postId: string;
};
