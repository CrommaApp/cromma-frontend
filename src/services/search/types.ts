export type AddRecentKeywordRequest = {
	content: string;
};
