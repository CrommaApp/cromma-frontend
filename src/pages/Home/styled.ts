import styled from 'styled-components';

export const MorePostsButton = styled.button`
	display: block;
	margin: auto;
	width: 25%;
	background-color: white;
	border: 2px solid #dddddd;
	border-radius: 8px;
	color: var(--font-color-dark-gray);
	padding: 2%;

	@media screen and (max-width: 768px) {
		width: 50%;
		padding: 3.5%;
	}
`;
