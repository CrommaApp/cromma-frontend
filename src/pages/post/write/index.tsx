import React from 'react';

const PostWrite = () => {
	return (
		<div>
			<>write</>
		</div>
	);
};

export default PostWrite;
