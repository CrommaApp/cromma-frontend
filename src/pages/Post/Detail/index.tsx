import React from 'react';
import PostDetailContent from '@components/features/Detail/DetailContent';
import { ContentWrapper } from '@components/shared/ContentWrapper/styled';
import useGetPost from '@hooks/useGetPost';
import useDeletePost from '@hooks/useDeletePost';

const PostDetail = () => {
	const [post, isMyPost] = useGetPost();

	const deletePost = useDeletePost(post.id);

	return (
		<>
			<ContentWrapper>
				<h1 className="a11y-hidden">게시글 전체 내용</h1>
				<PostDetailContent post={post} isMyPost={isMyPost} deletePost={deletePost} />
			</ContentWrapper>
		</>
	);
};

export default PostDetail;
