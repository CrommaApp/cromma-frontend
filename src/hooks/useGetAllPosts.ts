import React, { useEffect, useMemo, useState } from 'react';
import { useSetRecoilState } from 'recoil';
import PostService from '@services/post/post-service';
import { Post } from '@services/post/types';
import { errorStatusState } from '@stores/status';

type ReturnTypes = [Post[], boolean, boolean, () => Promise<void>];

const postService = new PostService();

const useGetAllPosts = (): ReturnTypes => {
	const [posts, setPosts] = useState<Post[]>([]);
	const [allPosts, setAllPosts] = useState<Post[]>([]);

	const isEmpty = useMemo(() => posts?.length === 0, [posts]);
	const isReachingEnd = useMemo(() => isEmpty || (posts && posts?.length < 10) || false, [isEmpty, posts]);

	const [isLoading, setIsLoading] = useState(false);

	const setErrorStatus = useSetRecoilState(errorStatusState);

	const getPosts = async (lastId: number) => {
		setIsLoading(true);
		try {
			const { data, statusCode } = await postService.getAllPosts(lastId);

			if (statusCode === 200 && data) {
				const newPosts = data.map((post) => {
					return {
						...post,
						createdAt: post.createdAt.slice(0, 10),
					};
				});
				setPosts(newPosts);

				return newPosts;
			}
		} catch (error: any) {
			setErrorStatus({
				errorMessage: '게시글을 불러오는데 실패했습니다.',
			});
		} finally {
			setIsLoading(false);
		}
	};

	const addToAllPosts = async () => {
		const newPosts = await getPosts(posts.pop()?.id || 0);

		if (newPosts) {
			setAllPosts((prev) => [...prev, ...newPosts]);
		}
	};

	const onClickMorePostsButton = async () => {
		if (isReachingEnd) return;

		addToAllPosts();
	};

	useEffect(() => {
		addToAllPosts();
	}, []);

	return [allPosts, isReachingEnd, isLoading, onClickMorePostsButton];
};

export default useGetAllPosts;
