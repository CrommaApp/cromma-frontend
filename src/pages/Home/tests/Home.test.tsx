import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { RecoilRoot } from 'recoil';
import Home from '@pages/Home';
import * as useGetAllPosts from '@hooks/useGetAllPosts';
import { Post } from '@services/post/types';

type ReturnType = [Post[], boolean, boolean, () => Promise<void>];

describe('Home Component', () => {
	const mockPosts: Post[] = [
		{
			id: 1,
			title: 'test1',
			content: 'lorem',
			createdAt: '2021-12-12',
			updatedAt: '2021-12-14',
			User: {
				userId: 'test@gmail.com',
			},
			UserId: 2,
		},
		{
			id: 2,
			title: 'test2',
			content: 'loremlorem',
			createdAt: '2021-12-10',
			updatedAt: '2021-12-12',
			User: {
				userId: 'test@gmail.com',
			},
			UserId: 2,
		},
	];

	const mockedOnClickMorePostsButton = jest.fn();

	const spyOnUseGetAllPosts = (returnData: ReturnType) => {
		const mockedUseGetAllPosts = jest.spyOn(useGetAllPosts, 'default');
		mockedUseGetAllPosts.mockImplementation(() => returnData);

		render(
			<RecoilRoot>
				<BrowserRouter>
					<Home />
				</BrowserRouter>
			</RecoilRoot>,
		);
	};

	it('Snapshot', () => {
		const component = renderer.create(
			<RecoilRoot>
				<BrowserRouter>
					<Home />
				</BrowserRouter>
			</RecoilRoot>,
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Show "no posts" if there is no post', () => {
		spyOnUseGetAllPosts([[], true, false, mockedOnClickMorePostsButton]);

		expect(
			screen.getByRole('heading', {
				name: '게시글이 없습니다.',
			}),
		).toBeInTheDocument();
	});

	it('Show "loading..." if posts is being imported', () => {
		spyOnUseGetAllPosts([[], false, true, mockedOnClickMorePostsButton]);

		expect(
			screen.getByRole('heading', {
				name: '불러오는 중...',
			}),
		).toBeInTheDocument();
	});

	it('Call onClickMorePostsButton() if click more posts button', () => {
		spyOnUseGetAllPosts([mockPosts, false, false, mockedOnClickMorePostsButton]);

		const morePostButton = screen.getByRole('button', {
			name: '더보기',
		});

		userEvent.click(morePostButton);

		expect(mockedOnClickMorePostsButton).toHaveBeenCalled();
	});
});
