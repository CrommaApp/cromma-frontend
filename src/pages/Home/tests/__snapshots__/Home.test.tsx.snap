// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component Snapshot 1`] = `
<section
  className="sc-eCImPb hmAVXe"
>
  <h1
    className="a11y-hidden"
  >
    전체 게시글 목록
  </h1>
  <div
    className="sc-jRQBWg fcMuTp"
  >
    <h1>
      게시글이 없습니다.
    </h1>
  </div>
</section>
`;
